File: app/repository/BaseRepository.java
package repository;

import io.ebean.Ebean;
import io.ebean.EbeanServer;
import io.ebean.Query;
import io.ebean.Transaction;
import models.BaseModel;
import play.db.ebean.EbeanConfig;

import javax.inject.Inject;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

import static java.util.concurrent.CompletableFuture.supplyAsync;

public abstract class BaseRepository<T extends BaseModel> {

    private final EbeanServer ebeanServer;
    private final DatabaseExecutionContext executionContext;

    @Inject
    public BaseRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext executionContext){
        this.ebeanServer = Ebean.getServer(ebeanConfig.defaultServer());
        this.executionContext = executionContext;
    }

    protected abstract Class<T> getModelClass();

    protected Query<T> query(){
        return ebeanServer.find(getModelClass());
    }

    protected <U> CompletionStage<U> get(Function<Query<T>, U> queryBuilder){
        return supplyAsync(() -> queryBuilder.apply(query()), executionContext);
    }

    protected CompletionStage<Optional<T>> get(Long id, Function<Query<T>, Query<T>> queryBuilder){
        return get(queryBuilder.andThen(query -> query.setId(id).findOneOrEmpty()));
    }

    protected CompletionStage<Optional<T>> updateModel(T data){
        return supplyAsync(() -> {
            Transaction txn = ebeanServer.beginTransaction();
            try {
                T saved = ebeanServer.find(getModelClass()).setId(data.id).findOne();
                if (saved != null) {
                    saved.update();
                    txn.commit();
                }
                return Optional.ofNullable(saved);
            } finally {
                txn.end();
            }
        }, executionContext);
    }

    public CompletionStage<Optional<T>> get(Long id) {
        return get(id, query->query);
    }

    public CompletionStage<T> insert(T data) {
        return supplyAsync(() -> {
            data.id = null;
            ebeanServer.insert(data);
            return data;
        }, executionContext);
    }

}

File: app/repository/BaseRepository.scala
package repository

import io.ebean.{Ebean, EbeanServer, Query}
import models.BaseModel
import play.db.ebean.EbeanConfig

import scala.collection.JavaConverters._
import scala.concurrent.Future

abstract class BaseRepository[T <: BaseModel] {
  protected def modelClass: Class[T]

  protected def ebeanConfig: EbeanConfig

  protected def executionContext: DatabaseExecutionContext

  protected def ebeanServer: EbeanServer = Ebean.getServer(ebeanConfig.defaultServer)

  protected def query: Query[T] = ebeanServer.find(modelClass)

  protected def execute[U](query: => U): Future[U] =
    Future[U] {
      query
    }(executionContext)

  protected def getList[U](query: => Query[T]): Future[List[T]] =
    execute {
      query.findList().asScala.toList
    }

  protected def getOneWhere(id: Long)(query: => Query[T]): Future[Option[T]] =
    execute {
      Option(query.setId(id).findOne())
    }

  protected def getOne(id: Long): Future[Option[T]] =
    execute {
      Option(query.setId(id).findOne())
    }

  def update(data: T): Future[Option[T]] =
    execute {
      ebeanServer.update(data)
    }.flatMap { _ => getOne(data.id) }


  def insert(data: T): Future[T] =
    execute {
      ebeanServer.insert(data)
      data
    }

  def fetchByRef(refId: String): Future[Option[T]] =
    execute {
      Option(query
        .select("*")
        .where().eq("refId", refId)
        .findOne())
    }

  def insertOrUpdateByRef(refId: String, data: T): Future[T] =
    fetchByRef(refId).flatMap{opt =>
      if (opt.isDefined) {
        data.id = opt.get.id
        this.update(data).map(_.get)
      } else {
        this.insert(data)
      }
    }
}
